lockbox-ios/Store/AutoLockStore.swift

```diff
@@ -0,0 +1,132 @@
+    private let routeActionHandler: RouteActionHandler
+
+    var timer: Timer?
+
+    init(dispatcher: Dispatcher = Dispatcher.shared,
+         userDefaults: UserDefaults = UserDefaults.standard,
+         settingActionHandler: SettingActionHandler = SettingActionHandler.shared,
+         routeActionHandler: RouteActionHandler = RouteActionHandler.shared) {
+
+        self.dispatcher = dispatcher
+        self.userDefaults = userDefaults
+        self.settingActionHandler = settingActionHandler
+        self.routeActionHandler = routeActionHandler
+
+        self.userDefaults.onLock.subscribe(onNext: { locked in
+            if locked {
+                self.stopTimer()
+            } else {
+                self.setupTimer()
+            }
+        }).disposed(by: self.disposeBag)
+
+        self.dispatcher.register
+            .filterByType(class: LifecycleAction.self)
+            .subscribe(self.lifecycleAction)
+            .disposed(by: self.disposeBag)
+
+    }
+
+    private func lifecycleAction(evt: Event<LifecycleAction>) {
+        if evt.element == .background {
+            self.userDefaults.onAutoLockTime
+                .take(1)
+                .subscribe(onNext: { (latest: AutoLockSetting) in
+                    if latest == AutoLockSetting.OnAppExit {
+                        self.lockApp()
+                    }
+                }).disposed(by: self.disposeBag)
+        }
+    }
+
+    private func setupTimer() {
+        self.userDefaults.onAutoLockTime
+            .take(1)
+            .subscribe(onNext: { (latest: AutoLockSetting) in
+                switch latest {
+                case .FiveMinutes:
+                    self.setTimer(seconds: 60 * 5)
+                case .OneHour:
+                    self.setTimer(seconds: 60 * 60)
+                case .OneMinute:
+                    self.setTimer(seconds: 60)
+                case .TwelveHours:
+                    self.setTimer(seconds: 60 * 60 * 12)
+                case .TwentyFourHours:
+                    self.setTimer(seconds: 60 * 60 * 24)
+                case .Never:
+                    return
+                case .OnAppExit:
+                    return
+                }
+            })
+            .disposed(by: self.disposeBag)
+    }
+
+    private func dataStoreAction(evt: Event<DataStoreAction>) {
+        guard let action = evt.element else { return }
+
+        switch action {
+        case .locked(let locked):
+            if locked {
+                self.stopTimer()
+            }
+        case .opened(let opened):
+            if opened {
+                self.setupTimer()
+            }
+        default:
+            return
+        }
+    }
+
+    private func setTimer(seconds: Int) {
+        let timerValue = self.userDefaults.double(forKey: SettingKey.autoLockTimerDate.rawValue)
+        if timerValue != 0 {
+            self.timer = Timer(fireAt: Date(timeIntervalSince1970: timerValue),
```

Comment:
I think there's an `RxCocoa` binding for `Timer`s so that you can avoid the `@objc` annotation.... might even save you from having to save the reference but I didn't investigate that much.

Reply:
Do you happen to have a link to that documentation? I am having trouble finding it.